src/sema/mutability.rs

```diff
@@ -237,39 +237,123 @@ fn read_expression(expr: &Expression, state: &mut StateCheck) -> bool {
+        Expression::StorageArrayLength { loc, .. } | Expression::StorageLoad { loc, .. } => {
             state.read(loc)
         }
-        Expression::Subscript(loc, _, ty, ..) if ty.is_contract_storage() => state.read(loc),
-        Expression::Builtin(loc, _, Builtin::GetAddress, _)
-        | Expression::Builtin(loc, _, Builtin::BlockNumber, _)
-        | Expression::Builtin(loc, _, Builtin::Timestamp, _)
-        | Expression::Builtin(loc, _, Builtin::ProgramId, _)
-        | Expression::Builtin(loc, _, Builtin::BlockCoinbase, _)
-        | Expression::Builtin(loc, _, Builtin::BlockDifficulty, _)
-        | Expression::Builtin(loc, _, Builtin::BlockHash, _)
-        | Expression::Builtin(loc, _, Builtin::Sender, _)
-        | Expression::Builtin(loc, _, Builtin::Origin, _)
-        | Expression::Builtin(loc, _, Builtin::Gasleft, _)
-        | Expression::Builtin(loc, _, Builtin::Gasprice, _)
-        | Expression::Builtin(loc, _, Builtin::GasLimit, _)
-        | Expression::Builtin(loc, _, Builtin::MinimumBalance, _)
-        | Expression::Builtin(loc, _, Builtin::Balance, _)
-        | Expression::Builtin(loc, _, Builtin::Random, _)
-        | Expression::Builtin(loc, _, Builtin::Accounts, _) => state.read(loc),
-        Expression::Builtin(loc, _, Builtin::PayableSend, _)
-        | Expression::Builtin(loc, _, Builtin::PayableTransfer, _)
-        | Expression::Builtin(loc, _, Builtin::SelfDestruct, _) => state.write(loc),
-        Expression::Builtin(loc, _, Builtin::ArrayPush, args)
-        | Expression::Builtin(loc, _, Builtin::ArrayPop, args)
-            if args[0].ty().is_contract_storage() =>
-        {
-            state.write(loc)
+        Expression::Subscript {
+            loc, array_ty: ty, ..
+        } if ty.is_contract_storage() => state.read(loc),
+        Expression::Builtin {
+            loc,
+            builtin: Builtin::GetAddress,
+            ..
         }
+        | Expression::Builtin {
+            loc,
+            builtin: Builtin::BlockNumber,
+            ..
+        }
+        | Expression::Builtin {
+            loc,
+            builtin: Builtin::Timestamp,
+            ..
+        }
+        | Expression::Builtin {
+            loc,
+            builtin: Builtin::ProgramId,
+            ..
+        }
+        | Expression::Builtin {
+            loc,
+            builtin: Builtin::BlockCoinbase,
+            ..
+        }
+        | Expression::Builtin {
+            loc,
+            builtin: Builtin::BlockDifficulty,
+            ..
+        }
+        | Expression::Builtin {
+            loc,
+            builtin: Builtin::BlockHash,
+            ..
+        }
+        | Expression::Builtin {
+            loc,
+            builtin: Builtin::Sender,
+            ..
+        }
+        | Expression::Builtin {
+            loc,
+            builtin: Builtin::Origin,
+            ..
+        }
+        | Expression::Builtin {
+            loc,
+            builtin: Builtin::Gasleft,
```

Comment:
Is it possible to match inside the builtin field?
Something on these lines:
```Rust
Expression::Builtin {
    loc,
    builtin: Builtin::Gasleft | Builtin::Origin // and so on
}
```

Reply:
Good point!